tensorflow_probability/python/distributions/empirical.py

```diff
@@ -0,0 +1,204 @@
+  dist.n_sample()
+
+  # Evaluate the pmf at 1, returning a scalar.
+  dist.prob(1.)
+
+  # Get 3 samples, returning a tensor with shape (3,).
+  dist.sample([3])
+  ```
+
+  """
+
+  def __init__(self,
+               params,
+               validate_args=False,
+               allow_nan_stats=True,
+               name="Empirical"):
+    """Initialize an `Empirical` random variable.
+    Args:
+      params: tf.Tensor representing a collection of samples. Its
+        outer (left-most) dimension determines the number of samples.
+      validate_args: Python `bool`, default `False`. When `True` distribution
+        parameters are checked for validity despite possibly degrading runtime
+        performance. When `False` invalid inputs may silently render incorrect
+        outputs.
+      allow_nan_stats: Python `bool`, default `True`. When `True`, statistics
+        (e.g., mean, mode, variance) use the value "`NaN`" to indicate the
+        result is undefined. When `False`, an exception is raised if one or
+        more of the statistic's batch members are undefined.
+      name: Python `str` name prefixed to Ops created by this class.
+    """
+    parameters = locals()
+    with tf.name_scope(name, values=[params]):
+      if (params is None):
+        raise ValueError("Params is missing.")
+      with tf.control_dependencies([]):
+        self._params = tf.convert_to_tensor(params, name="params")
+        try:
+          self._n_sample = tf.shape(self._params)[0]
+        except:  # scalar params
+          self._n_sample = tf.constant(1)
+
+    super(Empirical, self).__init__(
+        dtype=self._params.dtype,
+        reparameterization_type=reparameterization.FULLY_REPARAMETERIZED,
+        validate_args=validate_args,
+        allow_nan_stats=allow_nan_stats,
+        parameters=parameters,
+        graph_parents=[self._params, self._n_sample],
+        name=name)
+
+  @staticmethod
+  def _param_shapes(sample_shape):
+    return {"params": tf.convert_to_tensor(sample_shape, dtype=tf.int32)}
+
+  @property
+  def params(self):
+    """Distribution parameter."""
+    return self._params
+
+  @property
+  def n_sample(self):
+    """Number of samples."""
+    return self._n_sample
+
+  def _batch_shape_tensor(self):
+    return tf.constant([], dtype=tf.int32)
+
+  def _batch_shape(self):
+    return tf.TensorShape([])
+
+  def _event_shape_tensor(self):
+    return tf.shape(self.params)[1:]
+
+  def _event_shape(self):
+    return self.params.shape[1:]
+
+  def _mean(self):
+    return tf.reduce_mean(self.params, 0)
+
+  def _mode(self):
+    params = array_ops.reshape(self.params, [-1])
+    unique = tf.unique_with_counts(params)
+    return tf.gather(unique.y, tf.argmax(unique.count))
+
+  def _stddev(self):
+    # broadcasting n x shape - shape = n x shape
+    r = self.params - self.mean()
+    var = tf.cast(tf.reduce_mean(tf.square(r), 0), tf.float32)
+    return tf.sqrt(var)
+
+  def _variance(self):
```

Comment:
You can remove this method; the default `variance` implementation already does this if no variance is explicitly specified.

Reply:
Fixed.